File: 206-reverse-linked-list/206-reverse-linked-list.py
# Approach: Iterative

# Time complexity : O(n)
# Assume that n is the list's length, the time complexity is O(n)

# Space complexity : O(1)
    
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        curr = head
        while curr:
            next_temp = curr.next
            curr.next = prev
            prev = curr
            curr = next_temp
        return prev
        

File: 206-reverse-linked-list/206-reverse-linked-list.js
// Approach: Iterative
// Time complexity : O(n)
// Space complexity : O(1)

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function(head) {
    let prev = null;
    let curr = head;
    while (curr) {
        next_temp = curr.next;
        curr.next = prev;
        prev = curr;
        curr = next_temp;
    }
    return prev;   
};